### PR TITLE
[libc++] Implement LWG3430 disallow implicit conversion of the source arguments to `std::filesystem::path` when constructing `std::basic_*fstream`

### PR DESCRIPTION
**Pull request by yronglin**
Implement [LWG3430](https://wg21.link/LWG3430).

### CODE REVIEW

File: libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp
Comment (xiaoyang-sde) [inline libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp:34]: It seems that these code compiles even without this pull request because `fake_path` can't be converted to `filesystem::path`: https://godbolt.org/z/1dcsjdeEb

Comment (yronglin) [reply libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp]: Good catch! And thanks for your review! I've refined the test cases, also the title need update.
```
Note: the issue description is now a misnomer, as adding a string_view constructor is no longer being considered at this time.
```

**Review by mordante — CHANGES_REQUESTED** [review]
This [LWG](https://cplusplus.github.io/LWG/issue3430) issue does not what the title says. Based on the issue
```
 This means the simpler, more obvious code is slower and uses more memory:

    string_view sv = "foo.txt";
    fstream f1(sv); // bad
    fstream f2(string(sv)); // good

We should just allow passing a string_view directly, since it already compiles but doesn't do what anybody expects or wants. 
```
The accepted resolution avoids the `path` to `string_view` conversion, but doesn't add the `string_view` constructor overload. Based on https://godbolt.org/z/TqoWTxT75 the change would disallow using a `string_view` instead of making its usage more efficient. I wonder whether this change in behaviour is intended, especially since the title suggests differently and the LWG-issue doesn't change Annex C. @jwakely is the change in behaviour intended?

@yronglin let's put this patch on hold for now.

**Comment by jwakely** [comment]
> The accepted resolution avoids the `path` to `string_view` conversion, but doesn't add the `string_view` constructor overload. Based on https://godbolt.org/z/TqoWTxT75 the change would disallow using a `string_view` instead of making its usage more efficient. I wonder whether this change in behaviour is intended, especially since the title suggests differently and the LWG-issue doesn't change Annex C. @jwakely is the change in behaviour intended?

Yes, that was what LWG suggested and LEWG approved.

**Comment by mordante** [comment]
> > The accepted resolution avoids the `path` to `string_view` conversion, but doesn't add the `string_view` constructor overload. Based on https://godbolt.org/z/TqoWTxT75 the change would disallow using a `string_view` instead of making its usage more efficient. I wonder whether this change in behaviour is intended, especially since the title suggests differently and the LWG-issue doesn't change Annex C. @jwakely is the change in behaviour intended?
> 
> Yes, that was what LWG suggested and LEWG approved.

Thanks for the confirmation.

**Comment by yronglin** [comment]
> @yronglin Can you mention this change in behavior as a breaking change in the release notes.

Sure, added a release notes item.

File: libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp
Comment (yronglin) [reply libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp:25]: done

File: libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp
Comment (yronglin) [reply libcxx/test/std/input.output/file.streams/fstreams/fstream.cons/path.pass.cpp]: done

File: libcxx/include/fstream
Comment (yronglin) [reply libcxx/include/fstream:1108]: done

File: libcxx/include/fstream
Comment (yronglin) [reply libcxx/include/fstream]: done

File: libcxx/include/fstream
Comment (yronglin) [reply libcxx/include/fstream]: done

File: libcxx/docs/ReleaseNotes/19.rst
Comment (mordante) [inline libcxx/docs/ReleaseNotes/19.rst:100]: I think this is better placed under `Deprecations and Removals`. 

This text is technically correct. However the target audience of the release notes are typical developers and not people who read LWG issues and papers. So let's improve this text by explaining what it does. For example, 
```
This effectively removes the possibility to directly construct a ``std::basic_fstream`` from a
``std::basic_stringview`` or a C-string, instead you can construct a temporary ``std::basic_string``.
This change has been applied to C++17 and later.
```
Feel free to write this in your own words.

**Review by mordante — COMMENTED** [review]
This looks quite good. For now I'm happy to have this without a "old behaviour switch", but I've asked Louis whether he can test it internally to see whether this breaks things.

**Comment by ldionne** [comment]
@mordante Let me know about this and suggested it could be interesting to test on a large code base to see whether "removing" the `string_view` constructor creates some breakage. Please ping me once this is ready to try out and I can kick that off.

**Comment by yronglin** [comment]
> @mordante let me know about this and suggested it could be interesting to test on a large code base to see whether "removing" the `string_view` constructor creates some breakage. Please ping me once this is ready to try out and I can kick that off.

Sure!

File: libcxx/docs/ReleaseNotes/19.rst
Comment (yronglin) [reply libcxx/docs/ReleaseNotes/19.rst:105]: Thanks, I've update release notes wording.

**Comment by yronglin** [comment]
friendly ping~

**Comment by yronglin** [comment]
ping～

File: libcxx/docs/ReleaseNotes/19.rst
Comment (ldionne) [inline libcxx/docs/ReleaseNotes/19.rst:85]: You have trailing whitespace in here.

**Review by ldionne — COMMENTED** [review]
I'll take this for a run and let you know next week.

Edit: note to self 125553528

File: libcxx/docs/ReleaseNotes/19.rst
Comment (yronglin) [reply libcxx/docs/ReleaseNotes/19.rst]: Good catch!

**Comment by yronglin** [comment]
> I'll take this for a run and let you know next week.

Thanks for your review! Awaiting your good news.

**Comment by ldionne** [comment]
I am seeing a bit of breakage of the exact form we'd expect:

```
error: no matching constructor for initialization of 'std::ifstream' (aka 'basic_ifstream<char>')
In file included from [...]:
[...]/usr/include/c++/v1/__fwd/fstream.h:24:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::string_view' (aka 'basic_string_view<char>') to 'const ifstream' for 1st argument
```

However, this breakage seems surmountable (a few projects broke but not *that* many). It'll be slightly annoying but not a deal breaker.

After reading the LWG issue, my understanding is that the intended usage is to instead construct a `std::string` explicitly when passing it to the `std::fstream` so as to explicitly call out that we're creating a null-terminated string. Is that correct?

Pinging @llvm/libcxx-vendors for awareness but this LGTM.

**Comment by jwakely** [comment]
> After reading the LWG issue, my understanding is that the intended usage is to instead construct a `std::string` explicitly when passing it to the `std::fstream` so as to explicitly call out that we're creating a null-terminated string. Is that correct?

Yes. That was the intended usage, which was then broken by adding `filesystem::path` which allowed an implicit (but expensive) conversion.

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for testing the patch internally @ldionne!

LGTM modulo one minor issue. I'd like to have a quick look after that is resolved.

File: libcxx/test/std/input.output/file.streams/fstreams/ifstream.cons/path.pass.cpp
Comment (yronglin) [reply libcxx/test/std/input.output/file.streams/fstreams/ifstream.cons/path.pass.cpp]: Thanks, done! I've add a new macro `TEST_HAS_OPEN_WITH_WCHAR`.

**Comment by yronglin** [comment]
Seems the CI issue was not caused by this PR

**Review by EricWF — APPROVED** [review]
LGTM once the bots come back.

**Comment by yronglin** [comment]
CI has green and I merged this PR.

**Comment by mordante** [comment]
> CI has green and I merged this PR.

LGTM. Next time when somebody likes to review it before committing, please give them some time to look at the changes.

**Comment by yronglin** [comment]
> > CI has green and I merged this PR.
> 
> LGTM. Next time when somebody likes to review it before committing, please give them some time to look at the changes.

Thank you for the review and help, I will do this next time.

**Comment by nico** [comment]
This breaks existing code, e.g. this:

```
  std::fstream model_file(ml_package_dir()
                              .Append(kMlPackageDataDir)
                              .Append(kMlPackageModelFileName)
                              .value(),
                          std::ios::out | std::ios::binary);
```

`.value()` returns a `std::wstring`. (This is a Windows-only thing, which is probably why @ldionne didn't encounter it.)

Is the assumption that this will be rarely hit in practice?

<details>
<summary>full diag text</summary>

```
stderr:
../../services/webnn/coreml/graph_builder.cc(586,16): error: no matching constructor for initialization of 'std::fstream' (aka 'basic_fstream<char>')
  586 |   std::fstream model_file(ml_package_dir()
      |                ^          ~~~~~~~~~~~~~~~~
  587 |                               .Append(kMlPackageDataDir)
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~
  588 |                               .Append(kMlPackageModelFileName)
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  589 |                               .value(),
      |                               ~~~~~~~~~
  590 |                           std::ios::out | std::ios::binary);
      |                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../third_party/libc++/src/include\fstream(1411,34): note: candidate constructor not viable: no known conversion from 'const StringType' (aka 'const basic_string<wchar_t>') to 'const char *' for 1st argument
 1411 |   _LIBCPP_HIDE_FROM_ABI explicit basic_fstream(const char* __s,
      |                                  ^             ~~~~~~~~~~~~~~~
../../third_party/libc++/src/include\fstream(1414,34): note: candidate constructor not viable: no known conversion from 'const StringType' (aka 'const basic_string<wchar_t>') to 'const wchar_t *' for 1st argument
 1414 |   _LIBCPP_HIDE_FROM_ABI explicit basic_fstream(const wchar_t* __s,
      |                                  ^             ~~~~~~~~~~~~~~~~~~
../../third_party/libc++/src/include\fstream(1417,34): note: candidate constructor not viable: no known conversion from 'const basic_string<wchar_t>' to 'const basic_string<char>' for 1st argument
 1417 |   _LIBCPP_HIDE_FROM_ABI explicit basic_fstream(const string& __s,
      |                                  ^             ~~~~~~~~~~~~~~~~~
../../third_party/libc++/src/include\fstream(1422,74): note: candidate template ignored: requirement 'is_same_v<std::wstring, std::filesystem::path>' was not satisfied [with _Tp = StringType]
 1422 |   _LIBCPP_AVAILABILITY_FILESYSTEM_LIBRARY _LIBCPP_HIDE_FROM_ABI explicit basic_fstream(
      |                                                                          ^
../../third_party/libc++/src/include\fstream(1427,25): note: candidate constructor not viable: requires single argument '__rhs', but 2 arguments were provided
 1427 |   _LIBCPP_HIDE_FROM_ABI basic_fstream(basic_fstream&& __rhs);
      |                         ^             ~~~~~~~~~~~~~~~~~~~~~
../../third_party/libc++/src/include\__fwd/fstream.h(28,28): note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
   28 | class _LIBCPP_TEMPLATE_VIS basic_fstream;
      |                            ^~~~~~~~~~~~~
../../third_party/libc++/src/include\fstream(1410,25): note: candidate constructor not viable: requires 0 arguments, but 2 were provided
 1410 |   _LIBCPP_HIDE_FROM_ABI basic_fstream();
      |                         ^
```
</details>

(https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/coreml/graph_builder.cc;l=586?q=services%2Fwebnn%2Fcoreml%2Fgraph_builder.cc)

File: libcxx/include/fstream
Comment (nico) [inline libcxx/include/fstream:162]: Actually, https://en.cppreference.com/w/cpp/io/basic_fstream/basic_fstream mentions a `const std::filesystem::path::value_type*` overload which kinda happened to work before this change (since it'd construct a `::path` and call that ctor), but this change breaks that. Should we undo this change here until an explicit `const std::filesystem::path::value_type*` ctor exists?

Comment (jwakely) [reply libcxx/include/fstream:162]: That constructor already exists:

https://github.com/llvm/llvm-project/blob/a417b9b759960cc4d7b49e95ddc2ef129b5abe12/libcxx/include/fstream#L1413-L1415

It's definitely expected that you can't open an fstream with a `wstring`, the constructor aking a `filesystem::path` should not be used to convert "anything string-like, with any character type, in any encoding" into a file name with implicit allocations and character conversions.

Comment (jwakely) [reply libcxx/include/fstream:162]: If you want that conversion, you can still do it explicitly:

```c++
std::fstream model_file(filesystem::path(ml_package_dir()
                              .Append(kMlPackageDataDir)
                              .Append(kMlPackageModelFileName)
                              .value()),
                          std::ios::out | std::ios::binary);
```

Comment (nico) [reply libcxx/include/fstream:162]: Oh, I suppose I need to call `c_str()` on my wstring to get the wchar_t ctor. Let me try that; sorry for the noise.

(The point stands that this breaks existing code, though.)

Comment (mordante) [reply libcxx/include/fstream:162]: It's indeed correct this may break existing code. Louis tested this at Apple, based on the [results](
https://github.com/llvm/llvm-project/pull/85079#issuecomment-2032096059) we applied the possibly breaking change. Is the breakage for you worse?

Comment (nico) [reply libcxx/include/fstream:162]: It looks like it's just one file so far (*). I got a bit confused because the wchar_t* ctor is missing from the synopsis comment at the top. So all good.

*: But actually rolling this in is blocked on https://github.com/llvm/llvm-project/pull/86843#issuecomment-2051880399 , so not 100% sure yet. But very likely that's all.

**Pull request merged**
